src/IECoreScene/ShaderNetworkAlgo.cpp

```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
+					if( !splineValuesData )
+					{
+						throw IECore::Exception( "Internal failure in convertToOSLConventions - expandSpline did not create values" );
+					}
+
+					const std::vector< float > &splineValues = splineValuesData->readable();
+
+					adapterArraySize = splineValues.size();
+
+					if( adapterArraySize <= maxArrayInputAdapterSize )
+					{
+						for( unsigned int i = 0; i < splineValues.size(); i++ )
+						{
+							adapter->parameters()[ g_arrayInputNames[i] ] = new FloatData( splineValues[i] );
+						}
+					}
+				}
+
+				if( adapterArraySize > maxArrayInputAdapterSize )
+				{
+					throw IECore::Exception(
+						"Cannot handle input to " + connection.destination.shader.string() + "." +
+						destName + ". Expanded spline has " + std::to_string( adapterArraySize) +
+						" control points, but max input adapter size is " + std::to_string( maxArrayInputAdapterSize )
+					);
+				}
+
+				// TODO - does this need to be tagged to be reversible?
+				// ANSWER - it's a special node type, which users shouldn't be creating, so John pointed out
+				// we can probably just remove them all without needing to tag
+				//adapter->blindData()->writable()[ componentConnectionAdapterLabel() ] = g_trueData;
+
+				adapterHandle = network->addShader( s.first.string() + "InputArrayAdapter", std::move( adapter ) );
+				network->addConnection( ShaderNetwork::Connection(
+					{ adapterHandle, g_arrayOutputNames[ adapterArraySize ] },
+					{ s.first, splineValuesName }
+				) );
+
+				inserted.first->second.first = adapterHandle;
+				inserted.first->second.second = knotOffset;
+			}
+			else
+			{
+				adapterHandle = inserted.first->second.first;
+				knotOffset = inserted.first->second.second;
+			}
+
+			// Default is guaranteed not to be used, since from_chars must either set it or set the error code,
+			// but apparently the compiler doesn't know that
+			int elementId = -1;
+			const char *idStart = &destName[splineElementMatch.position(2)];
+			const char *idEnd = idStart + splineElementMatch.length(2);
+			auto elementIdResult = std::from_chars( idStart, idEnd, elementId );
+
+			if( elementIdResult.ec == std::errc::invalid_argument || elementIdResult.ptr != idEnd )
+			{
+				throw IECore::Exception( "Invalid spline point index " + splineElementMatch[2] );
+			}
+
+			InternedString destinationName = g_arrayInputNames[elementId + knotOffset];
+			if( splineElementMatch.length( 3 ) )
+			{
+				std::string nameWithSuffix;
+				nameWithSuffix.reserve( destinationName.string().size() + splineElementMatch.length(3) );
+				nameWithSuffix.append( destinationName.string() );
+				nameWithSuffix.append( &destName[splineElementMatch.position(3)], splineElementMatch.length(3) );
+
+				destinationName = nameWithSuffix;
+			}
+
+			network->removeConnection( connection );
+			network->addConnection( { connection.source, { adapterHandle, destinationName } } );
```

Comment:
Do we need to add multiple connections here sometimes, to account for endpoint duplication?

Reply:
The endPoint duplication that affects the resulting spline shape is done already when we create a SplineData.  The only end point duplication done in IECoreScene::shaderNetworkAlgo is just to account for the extra array positions demanded by OSL for linear and constant.  These values are not actually used - OSL will always treat a 4 component linear array `[ a, b, c, d ]` as a line from `b -> c`, and ignore the values of `a` and `d`.  It is a bit tidier looking to set them to the end values when translating the values, but it doesn't seem worthwhile to create extra connections for the sake of these unused values?

Reply:
Ah, cool, that makes sense. Maybe it's a bit misleading to be providing values for the duplicates but not connections though? Maybe just a comment that OSL will ignore the values, so we don't set or connect them?